### ISSUE TITLE
games that need nvram bootstraps in order to boot in playable condition

### ISSUE THREAD

**Issue by ghost** [libretro/mame2003-plus-libretro#134]
`zookeep` starts in Service Mode on first boot. The user has to simply restart the game to bypass. @markwkidd instructed me to open an issue about this.

**Comment by markwkidd** [comment]
Thanks.

I'd like to try the nvram bootstrap approach on this.

Until a good solution for de-hard-coding `F2` for service mode is available I want to try to bootstrap anything that requires a real F2 and would therefore be hard for someone with a SNES pad.

**Comment by arcadez** [comment]
Try Qix im sure you had to do the same kinda thing for that game as well

**Comment by markwkidd** [comment]
I would like to get a "nvram bootrap" list going

Also, it's pretty straightforward to add these bootstraps. I think that
other folks can add these nvram patches as well although it will be my
pleasure to continue :)

**Comment by SapphireTactician** [comment]
Actually, with Zoo Keeper, Service Mode is mapped to F1 by default. It's service mode and service mode controls are bizarre.

![2018-04-24--161252_1600x900](https://user-images.githubusercontent.com/24427487/39211690-714bb27a-47da-11e8-996f-2f677f51a5d6.png)

`Test Advance` is mapped to `F1` by default and is actually the service button for this game. Pressing it advances from the Location Name screen to a series of service menus. If pressed during gameplay, you'll enter Service Mode.

`Test Next Line` is mapped to `F2` by default, and is used to navigate and will wrap around back to the top after it reaches the bottom of a page. On the location screen, it advances the text input to the next space.

`Test Slew Up` is mapped to `F5` by default, and is used to cycle upward through letters on the location screen. It is also used to cycle forward through available options for a setting.

`Test Slew Down` is mapped to `F6` by default, and is used to cycle downward through letters on the location screen. It is also used to cycle backwards through available options for a setting.

-----

@arcadez Will check out Qix momentarily.

**Comment by markwkidd** [comment]
Ugh. Some of that Zoo Keeper stuff should be left to future generations. Because it should be looked at... maybe future MAME already has.

Anyway, as long as it's possible to get Zoo Keeper booting into fun, playable state with a SNES controller I'll be happy.

**Comment by SapphireTactician** [comment]
@arcadez @markwkidd Yes, can confirm that `qix` has the exact same boot sequence and service mode as Zoo Keeper.

**Comment by SapphireTactician** [comment]
We can add `Sinistar` to this list too. Seems to also start in service mode and require a restart.

**Comment by markwkidd** [comment]
Here is a list of all romsets that might have this issue:

`GAME( 1981, qix,      0,        qix,     qix,      0,        ROT270, "Taito America Corporation", "Qix (set 1)" )
GAME( 1981, qixa,     qix,      qix,     qix,      0,        ROT270, "Taito America Corporation", "Qix (set 2)" )
GAME( 1981, qixb,     qix,      qix,     qix,      0,        ROT270, "Taito America Corporation", "Qix (set 3)" )
GAME( 1981, qix2,     qix,      qix,     qix,      0,        ROT270, "Taito America Corporation", "Qix II (Tournament)" )
GAME( 1981, sdungeon, 0,        mcu,     sdungeon, 0,        ROT270, "Taito America Corporation", "Space Dungeon" )
GAME( 1982, elecyoyo, 0,        mcu,     elecyoyo, 0,        ROT270, "Taito America Corporation", "The Electric Yo-Yo (set 1)" )
GAME( 1982, elecyoy2, elecyoyo, mcu,     elecyoyo, 0,        ROT270, "Taito America Corporation", "The Electric Yo-Yo (set 2)" )
GAME( 1982, kram,     0,        mcu,     kram,     kram,     ROT0,   "Taito America Corporation", "Kram (set 1)" )
GAME( 1982, kram2,    kram,     mcu,     kram,     kram,     ROT0,   "Taito America Corporation", "Kram (set 2)" )
GAMEX(1982, kram3,    kram,     qix,     kram,     kram,     ROT0,   "Taito America Corporation", "Kram (encrypted)", GAME_UNEMULATED_PROTECTION )
GAME( 1982, zookeep,  0,        zookeep, zookeep,  zookeep,  ROT0,   "Taito America Corporation", "Zoo Keeper (set 1)" )
GAME( 1982, zookeep2, zookeep,  zookeep, zookeep,  zookeep,  ROT0,   "Taito America Corporation", "Zoo Keeper (set 2)" )
GAME( 1982, zookeep3, zookeep,  zookeep, zookeep,  zookeep,  ROT0,   "Taito America Corporation", "Zoo Keeper (set 3)" )
GAME( 1982, slither,  0,        slither, slither,  slither,  ROT270, "Century II",                "Slither (set 1)" )
GAME( 1982, slithera, slither,  slither, slither,  slither,  ROT270, "Century II",                "Slither (set 2)" )
GAME( 1984, complexx, 0,        qix,     complexx, 0,        ROT270, "Taito America Corporation", "Complex X" )`

**Comment by markwkidd** [comment]
Howdy @SapphireTactician . I have just pushed bootstraps for `zookeep`, `zookeep2`, `zookeep3`, `qix`, `qix2`, `qixa`, and `qixb`

Based on some very cursory testing on my system all of these romsets now boot past the service issue.

From your experience do you know if any of the other `qix` driver romsets listed in my previous message have service button/bootstrap issues?

**Comment by markwkidd** [comment]
I have just pushed bootstraps for `sinistar`, `sinista1`, and `sinista2`

That brings me to the end of all of the games that people have mentioned, other than the Midway/Mortal Kombat volume issues which we have a separate issue for

**Comment by SapphireTactician** [comment]
I haven't played any of the listed games other than Qix and Zoo Keeper. I'll look into as soon as things calm down IRL. Unfortunately, I'm really busy and behind schedule atm.

**Comment by markwkidd** [comment]
per arcadez: Charlie Ninja needs a bootstrap

**Comment by markwkidd** [comment]
TODO: look through the retropie mame2003 compatibility list for clues about other games needing nvram bootstrap https://docs.google.com/spreadsheets/d/1LP1MELCvcxu7TfiowF_0ZuvRVEMqlfQyTVetnOJvuJc/edit?usp=sharing

**Comment by arcadez** [comment]
Got another one for ya Defender all versions

**Comment by markwkidd** [comment]
I just added a bootstrap for `rungunu` -- I don't know why I didn't think to check it when I did `rungun`.

I took that opportunity to write a doc about the process in case anyone is curious about it works: https://github.com/libretro/mame2003-plus-libretro/wiki/Adding-an-NVRAM-Bootstrap

I will work on Charlie Ninja and the various Defender titles as soon as I get a chance.

**Comment by arcadez** [comment]
Narc as i recall needs a reset before you can get in game on first run boot.

**Comment by markwkidd** [comment]
bootstraps added for narc and narc3!

**Comment by arcadez** [comment]
Last couple i could find while checking out the compat list...........

Joust 1 and Joust 2

**Comment by Wilstorm** [comment]
Would Food Fight (foodf.zip) be this same issue? Basically I had turn on service mode, exit the game, start the game, reset within the options, turn off service mode, exit the game and restart. Then everything was right as rain.

**Comment by Wilstorm** [comment]
Here's a list of games I found that need either dip switch settings or the nvram tweaked to enable all players. By default these games have only 2 players enabled. I added .cfg or .nv after the name to signify what needs modified. I also listed all the clones but haven't tested them all. I usually use the US, European or World/International versions.

**Captain America and the Avengers (.cfg)**
captaven.zip
captavna.zip
captavne.zip
captavnj.zip
captavnu.zip
captavuu.zip

**Captain Commando (.cfg)**
captcomj.zip
captcomm.zip
captcomu.zip

**Dungeons & Dragons: Shadow over Mystara (.nv)**
ddsom.zip
ddsoma.zip
ddsomj.zip
ddsomjr1.zip
ddsomr1.zip
ddsomu.zip
ddsomur1.zip

**Karate Blazers (.cfg)**
karatblu
karatblz

**Mystic Warriors (.cfg)**
mystwarr.zi
mystwaru.zip

**The Combatribes (.cfg)**
ctribe.zip
ctribeb.zip

**Comment by etatto** [comment]
`bucky` might be another candidat: The game needs to be reseted with the Test switch (F2) down to pass the EEPROM N2 BAD message when first run. moomesa also needs this "workaround" (same driver).

**Comment by Wilstorm** [comment]
@markwkidd - I was doing some cleanup of old NVRAM files and I noticed a few bootstrap issues listed below. The one listed above by etatto seems to boot fine without issue.

Food Fight (foodf) - sits at service menu. 2nd boot does **not** start the game. You have to reset in service mode to get a working NVRAM file.
Joust (joust) - sits at factory settings restored. 2nd boot starts the game.
Robotron: 2084 (robotron) - sits at factory settings restored. 2nd boot starts the game.

_Games that tested good:_
Defender (defender)
Narc (narc)
Qix (qix)
Sinistar (sinistar)

**Comment by Wilstorm** [comment]
@markwkidd - I know you seem to have some issues with email getting flagged as spam but I thought I would post here one more time just in case you missed it. If not I guess people will figure it out eventually.

Food Fight and Joust are listed above so I don't know if an update broke them but surprisingly Robotron slipped through the cracks.

I don't think it's to much of an issue for the PC and Pi folks as they can be generated on the fly but I think you had mentioned some other platforms that require the bootstrap to play.

Atari's Food Fight might need an embedded bootstrap because the simple double reboot doesn't seem to fix that game.

Anyway Happy Halloween fellas and have a spooktacular day. I have a cute young lady that's going to be a bee that is super excited to go trick-or-treating tonight! :)

**Comment by Wilstorm** [comment]
@markwkidd - Here's the bootstraps for Food Fight, Joust, Joust 2 & Robotron, that includes the clones also in zip format. There's a readme with the ROM and MD5 hash right behind it. You were right some match and some don't. https://ufile.io/6hn2u

```
foodf		6926e04d538ce50aa81c0b8d00f677a8
foodf2		6926e04d538ce50aa81c0b8d00f677a8
foodfc		7988da261cd1be78d02d2cb888144efb

joust		76773c65e227c2a218fe3b0e45199c90
joustr		76773c65e227c2a218fe3b0e45199c90
joustwr		76773c65e227c2a218fe3b0e45199c90

joust2		15c8135db0cb597cbb3b5c8813a52d34

robotron	59c8c95abfad09347785f5fef258afcb
robotryo	67d89e992b95cf08d6ef757dd738aab6
```

**Comment by Wilstorm** [comment]
It was random games I came across. I do keep a list as I find them. There's others that give a nvram error and you only need to press a key to continue I have listed.

**Comment by markwkidd** [comment]
If it only takes pressing a regular button like a user would have already mapped if they're playing with a SNES controller or an arcade panel, then I would say that no nvram bootstrap is needed.

If they need to press something keyboard-only then that would maybe make my list.

**Comment by grant2258** [comment]
we could map a 3 key combo to get to the service like start + sel + a as default well i would imagine you could. there is no way you would accidently press them 3 together

**Comment by Wilstorm** [comment]
Yeah they are a simple key press to bypass

**Comment by markwkidd** [comment]
My purpose is to create a good user experience. If it's reasonably clear what the user needs to do and it's reasonably in their power to do so, no problem.

As far as games that need `nv` or `cfg` to add support for all players. I'm inclined to at least start adding accepting `cfg` and `nv` submissions to this repo as we are with the backdrop artwork. Later on if it seems good to have them automatically bootstrap into place, then half of the work is already done by gathering the files themselves.

@Wilstorm if I set up folders in `/metadata/` for these "player enable" bootstraps, would you be willing to contribute one or two `cfg`s and/or `nv`s to get the ball rolling?

**Comment by Wilstorm** [comment]
@markwkidd - Yeah that sounds good. Then I can learn how to do it as I come across them.

**Comment by Wilstorm** [comment]
> bucky might be another candidat: The game needs to be reseted with the Test switch (F2) down to pass the EEPROM N2 BAD message when first run. moomesa also needs this "workaround" (same driver).

I tested bucky, buckyua, moo, moobl and mooua all seemed to boot fine with no issues. It does have to initialize the EEPROM on the first boot but no user intervention is required.

**Comment by mahoneyt944** [comment]
Both Food fight games need the joystick calibration included for both upright and cocktail in the nv boot strap. You can get this by 1st loading the current boot strap then enabling cocktail in the service menu. Next calibrate both player 1 and player 2 (you'll need two controllers for this). This gives you the Cocktail food fight boot strap, then with the same file just turn cocktail back off in the menu to get the upright food fight bootstrap

**Comment by mahoneyt944** [comment]
@markwkidd @twinaphex

Created PR #840 to include a joystick calibration in the bootstrap for food fight, upright and cocktail.

**Comment by Wilstorm** [comment]
@mahoneyt944 - Thanks for the bootstrap. I was bogged down with switch replacements and then we had a switch failure last night, right in the middle of the stack no less. I didn't get out of here until almost midnight. I grabbed a copy of your instructions off the RP forums. I have a few days off coming up soon here and look forward to testing it! :)

**Comment by mahoneyt944** [comment]
@Wilstorm 

With this new boot strap it should eliminate the setup hassles. Just turn the game on and it will work as it should (face right when stopped). Then turn on the live center hack in the dip menu and adjust the debounce delay to use the new hack if desired.

**Comment by mahoneyt944** [comment]
@Wilstorm 

The new bootstrap is merged. Updated the instructions here. https://retropie.org.uk/forum/topic/25593/food-fight-joystick-troubles-solved/119

**Comment by Wilstorm** [comment]
@mahoneyt944 - This works great, he does face the direction he was moving when stopped. Nice job! :)

A quick question, with the new bootstrap we don't need to go to the dip switch setup or calibrate the joystick with the new bootstrap?

I use a XBOX 360 with the RA option "dpad to analog" (basically it allows you to use the left analog stick to play games). 

I did go through the setup and just held it up, down, left, right for 5 seconds each. I wasn't quite sure if that's what was needed as it said hold in each direction so I figured the basic directions but the original was an analog I think?

Basically what I was doing while playing. Go into the settings leaving live center on and then just adjusting the debounce and exit. I couldn't really tell a difference between 0 and 10 though. If it's subtle increments that's probably why. 

Either way (0 or 10) it seemed to work the same for me but all were good. I would move in one direction and then let up on the stick and he seemed to take maybe a half step or so whether it was 0, 2 or 10.

**Comment by mahoneyt944** [comment]
@Wilstorm

The update to the boot strap includes the calibration now so you don't need to perform one manually. Just delete the old foodf.nv file (bootstrap) and let the core automatically create a new one at game launch. 

This will allow the game to work as it does in all other cores (stops always facing right). To utilize the new control hack, turn live center on in the dip menu and play the game.

Debounce delay is important to set properly. Typically you'll get a setting of about 2-4. Here's tips on setting debounce:

**Debounce Delay Tips**
On a sanwa jlf 8 way digital joystick with 8 way restrictor in place a delay of 2 works well for me. The delay has to be set based on user preference and controller type though so adjust until it feels right. I typically test to make sure I can stop on the diagonals without any trouble. Then test quick directional changes left to right to see that it lands where I want. If it doesn't the delay may be set too high.

On a analog joystick I have a dead zone of 15 set in the core options but this can be adjusted to preference as well. Delay can be set at 0 for analog but I find it sets stops too fast for some of my controllers because it's setting stops as your releasing the joystick, so a little delay here helps too.

**Comment by Wilstorm** [comment]
@mahoneyt944 - Ah, thanks for the information that makes more sense. I am using the analog stick on a 360 controller and it feels good at 0 as well as 10. I just did a retest and I can "quick stop" and turn with all directions including diagonals with no issues with any debounce setting. 

I'm not sure what the default dead zone is on m3plus but I've never had to tweak it as most games feel pretty good. It does play well for me and I can chuck pies with the best of them. It's little tweaks like these that are the spit and shine. Thanks a bunch for this update! 

I will let you know if I ever get a hold of a digital stick for testing. Would it benefit at all to trying to set a comprise default of like 2 or 3 vs. 0 and have people tweak from there if some posts it's to fast or to slow? I like when something works out of the box and then you can tweak it from there.

Hopefully others will weigh in as the steps are very quick and easy to setup at a minute or so and it's actually pretty fun playing...errr....testing! ;)

**Comment by Wilstorm** [comment]
@mahoneyt944 - My deadzone is set to 20 which seems to play well with Food Fight.

**Comment by Wilstorm** [comment]
@mahoneyt944 - One last thing as I'm not sure if it matters or not but it doesn't seem to write the setting out to ```/opt/retropie/configs/all/retroarch-core-options.cfg``` whether it's enabled or disabled. Pretty much every option shows up in core options after you run a game and can be manually modified.

**Comment by Wilstorm** [comment]
I meant the ```Center joystick axis for digital controls```, sorry my brain is baked for the day...

**Comment by mahoneyt944** [comment]
> I meant the `Center joystick axis for digital controls`, sorry my brain is baked for the day...

It is in the core options. The option is:
`mame2003-plus_digital_analog = "enabled"`

As far as a middle ground setting I decided to go with 0 because the game is natively analog. And I didn't want a hacked setting to be native.

**Comment by Wilstorm** [comment]
@mahoneyt944 - Ah, ok, yep I see it in there now. Definitely could be renamed closer to the core option name but it's all good.  I agree it's nice to have the option to play it with digital controls.

**Comment by mahoneyt944** [comment]
@Wilstorm renamed #849

**Comment by mahoneyt944** [comment]
@ghost I think this can be closed now. All these titles are included in bootstraps now. Thank you.

**Comment by Wilstorm** [comment]
@mahoneyt944 - "Ghost" is what Github names an account once it's been deleted so this one will need to be closed by someone with permissions to do it.

Also I had a list of games that don't have the max players set by default. For example a game might be capable of 4 players but only 2 are enabled by default. Some are set through the dip switch settings and some through nvram. So it's a combination of .cfg and .nvram files.

Mark had mentioned maybe a metadata folder to dump the files for folks to use optional configs or we could tweak the default dips (.cfg files) and maybe bootstrap the others (.nvram files). If you don't think it's worth the time that's ok too but just wanted to mention since you've been a roll! :)

My current list of games that are not set to max players:

```
Captain America and the Avengers (.cfg)
captaven.zip
captavna.zip
captavne.zip
captavnj.zip
captavnu.zip
captavuu.zip

Captain Commando (.cfg)
captcomj.zip
captcomm.zip
captcomu.zip

Dungeons & Dragons: Shadow over Mystara (.nv)
ddsom.zip
ddsoma.zip
ddsomj.zip
ddsomjr1.zip
ddsomr1.zip
ddsomu.zip
ddsomur1.zip

Karate Blazers (.cfg)
karatblu.zip
karatblz.zip

Mystic Warriors (.cfg)
mystwarr.zip
mystwaru.zip

The Combatribes (.cfg)
ctribe.zip
ctribeb.zip
```

**Comment by Wilstorm** [comment]
I mentioned here as you can read more about it further up in this thread.

**Comment by mahoneyt944** [comment]
@Wilstorm I'm all for updating the cfg/ bootstraps to set the max players but that should be opened in a new dedicated issue. This issue was more for games that won't boot at all without a bootstrap do to eeprom or needing reset.

**Comment by Wilstorm** [comment]
Yeah, I know it was for non-booting games. We had some "cross talk" back in the day where one issue would lead to another similar issue. It was 2-3 years ago there was so many PR's going in to place you could hardly keep up with the changes. It was a happening core! ;)

I'll post the list I have over in the other thread and if someone feels like taking it on at some point if there's enough interest. I'm sure there are more games but these in my list of 330 or so "must haves" that I noticed needed a tweak to get all the players enabled. I listed the parent and clones even though I mainly stick to US and European versions.

The other challenge is some have single coin chutes or a coin chute per player. Depending on your setup you may have one or several coin buttons on your panel or if you use controllers it nice for each player to add their own coins. Basically there may be a few things that need tweaked.

**Comment by mahoneyt944** [comment]
Moved to #875
